modal/_utils/async_utils.py

```diff
@@ -544,3 +545,116 @@ async def producer(producer_id: int, iterable: Union[AsyncIterable[T], Iterable[
 
 async def callable_to_agen(awaitable: Callable[[], Awaitable[T]]) -> AsyncGenerator[T, None]:
     yield await awaitable()
+
+
+async def async_map(
+    input: Union[AsyncIterable[T], Iterable[T]],
+    async_mapper_func: Callable[[T], Awaitable[V]],
+    concurrency: int,
+) -> AsyncGenerator[V, None]:
+    input_queue: asyncio.Queue[Tuple[str, Union[T, None]]] = asyncio.Queue()
+    output_queue: asyncio.Queue[Tuple[str, Union[V, Exception]]] = asyncio.Queue()
+    output_event = asyncio.Event()
+
+    async def producer():
+        async for item in sync_or_async_iter(input):
+            await input_queue.put(("value", item))
+        await input_queue.put(("stop", None))
+
+    async def worker():
+        while True:
+            try:
+                event_type, item = await input_queue.get()
+                if event_type == "stop":
+                    break
+                if event_type == "value":
+                    if asyncio.iscoroutinefunction(async_mapper_func):
+                        result = await async_mapper_func(item)
+                    else:
+                        result = typing.cast(V, async_mapper_func(item))
+                    await output_queue.put(("value", result))
+            except Exception as e:
+                await output_queue.put(("exception", e))
+            finally:
+                input_queue.task_done()
+
+    output_event_task = asyncio.create_task(output_event.wait())
+    producer_task = asyncio.create_task(producer())
+    worker_tasks = [asyncio.create_task(worker()) for _ in range(concurrency)]
+
+    async def complete_map():
+        await producer_task
+        await input_queue.join()
+
+    complete_map_task = asyncio.create_task(complete_map())
+
+    all_tasks = [output_event_task, producer_task, *worker_tasks, complete_map_task]
+
+    try:
+        while True:
+            done, _ = await asyncio.wait(
+                all_tasks,
+                return_when=asyncio.FIRST_COMPLETED,
+            )
+
+            finished_workers = done & set(worker_tasks)
+            for finished_worker in finished_workers:
+                await finished_worker
+
+            if complete_map_task.done():
+                while not output_queue.empty():
+                    event_type, item = await output_queue.get()
+                    if event_type == "value":
+                        yield typing.cast(V, item)
+                    elif event_type == "exception":
+                        raise typing.cast(Exception, item)
+                    else:
+                        raise Exception("Unknown event type: " + event_type)
+                break
```

Comment:
should add a `await complete_map_task` here before the break - in case it has exceptions that needs to be surfaced, similar to the worker tasks. Otherwise those exceptions would never get surfaced, since the `finally` clause uses `return_exceptions=True`